File: backend/internal/accounting_export/xero.go
package accounting_export

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/google/uuid"
)

// ─── Xero OAuth2 + API configuration ─────────────────────────────────────────

// XeroConfig holds OAuth2 credentials and endpoints. All values are injected
// from environment variables (XERO_CLIENT_ID, XERO_CLIENT_SECRET,
// XERO_REDIRECT_URL). No defaults for secrets; missing values produce an
// error at construction time.
type XeroConfig struct {
	ClientID     string
	ClientSecret string
	RedirectURL  string
	// Scopes is the list of Xero OAuth2 scopes to request.
	// Defaults to the recommended set when nil.
	Scopes []string
}

// DefaultXeroScopes are the Xero API scopes required for contacts +
// bank transactions. "offline_access" is required for refresh tokens.
var DefaultXeroScopes = []string{
	"openid",
	"profile",
	"email",
	"accounting.contacts",
	"accounting.transactions",
	"offline_access",
}

// xeroTokenResponse is the JSON body returned by the Xero token endpoint.
type xeroTokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in"` // seconds
	IDToken      string `json:"id_token"`
}

// xeroConnectionsResponse is the JSON from GET /connections.
type xeroConnectionsResponse []struct {
	TenantID   string `json:"tenantId"`
	TenantType string `json:"tenantType"`
	TenantName string `json:"tenantName"`
}

// ─── XeroHTTPClient interface (allows mock in tests) ─────────────────────────

// XeroHTTPClient is the subset of *http.Client that XeroProvider uses.
// Swap in a test double via NewXeroProviderWithClient to avoid live network
// calls in unit tests. Gate real integration tests behind build tag `integration`.
type XeroHTTPClient interface {
	Do(req *http.Request) (*http.Response, error)
}

// ─── Token encryption ─────────────────────────────────────────────────────────

// TokenCipher handles symmetric encryption of OAuth tokens before storage in
// oauth_grants. The real implementation uses AES-256-GCM with a key derived
// from APP_SECRET. Keeping it as an interface lets tests inject a no-op.
type TokenCipher interface {
	Encrypt(plaintext []byte) ([]byte, error)
	Decrypt(ciphertext []byte) ([]byte, error)
}

// PlaintextCipher is a no-op TokenCipher used in development/tests.
// In production replace with an AES-GCM implementation keyed from APP_SECRET.
type PlaintextCipher struct{}

func (PlaintextCipher) Encrypt(b []byte) ([]byte, error) { return b, nil }
func (PlaintextCipher) Decrypt(b []byte) ([]byte, error) { return b, nil }

// ─── XeroProvider ─────────────────────────────────────────────────────────────

// XeroProvider implements Provider for Xero. It is safe for concurrent use.
type XeroProvider struct {
	cfg    XeroConfig
	store  *Store
	client XeroHTTPClient
	cipher TokenCipher
}

// NewXeroProvider constructs a XeroProvider using the standard http.Client and
// PlaintextCipher. For production use swap cipher for a real AES-GCM cipher.
func NewXeroProvider(cfg XeroConfig, store *Store) (*XeroProvider, error) {
	if cfg.ClientID == "" || cfg.ClientSecret == "" || cfg.RedirectURL == "" {
		return nil, errors.New("xero: XERO_CLIENT_ID, XERO_CLIENT_SECRET and XERO_REDIRECT_URL are required")
	}
	if len(cfg.Scopes) == 0 {
		cfg.Scopes = DefaultXeroScopes
	}
	return &XeroProvider{
		cfg:    cfg,
		store:  store,
		client: &http.Client{Timeout: 15 * time.Second},
		cipher: PlaintextCipher{},
	}, nil
}

// NewXeroProviderWithClient constructs a XeroProvider with an injected HTTP
// client and cipher. Used in tests to avoid live network calls.
func NewXeroProviderWithClient(cfg XeroConfig, store *Store, client XeroHTTPClient, cipher TokenCipher) (*XeroProvider, error) {
	if cfg.ClientID == "" || cfg.ClientSecret == "" || cfg.RedirectURL == "" {
		return nil, errors.New("xero: XERO_CLIENT_ID, XERO_CLIENT_SECRET and XERO_REDIRECT_URL are required")
	}
	if len(cfg.Scopes) == 0 {
		cfg.Scopes = DefaultXeroScopes
	}
	return &XeroProvider{
		cfg:    cfg,
		store:  store,
		client: client,
		cipher: cipher,
	}, nil
}

// Name implements Provider.
func (p *XeroProvider) Name() string { return "xero" }

// AuthURL implements Provider. Returns the Xero consent URL.
// state must be a cryptographically random, session-bound nonce validated
// in the callback handler.
func (p *XeroProvider) AuthURL(orgID uuid.UUID, state string) string {
	params := url.Values{}
	params.Set("response_type", "code")
	params.Set("client_id", p.cfg.ClientID)
	params.Set("redirect_uri", p.cfg.RedirectURL)
	params.Set("scope", strings.Join(p.cfg.Scopes, " "))
	params.Set("state", state)
	return "https://login.xero.com/identity/connect/authorize?" + params.Encode()
}

// ExchangeCode implements Provider. Exchanges an authorisation code for tokens
// and stores them encrypted in oauth_grants. It also calls GET /connections to
// fetch the Xero tenant ID (stored as account_email for display purposes).
func (p *XeroProvider) ExchangeCode(ctx context.Context, orgID, userID uuid.UUID, code string) (string, error) {
	tok, err := p.tokenRequest(ctx, url.Values{
		"grant_type":   {"authorization_code"},
		"code":         {code},
		"redirect_uri": {p.cfg.RedirectURL},
	})
	if err != nil {
		return "", fmt.Errorf("xero: exchange code: %w", err)
	}

	tenantID, err := p.fetchTenantID(ctx, tok.AccessToken)
	if err != nil {
		return "", fmt.Errorf("xero: fetch tenant id: %w", err)
	}

	if err := p.storeTokens(ctx, orgID, userID, tenantID, tok); err != nil {
		return "", fmt.Errorf("xero: store tokens: %w", err)
	}
	return tenantID, nil
}

// RefreshToken implements Provider. Fetches a new access token using the stored
// refresh token and persists the updated pair to oauth_grants.
func (p *XeroProvider) RefreshToken(ctx context.Context, orgID uuid.UUID) error {
	grant, err := p.store.GetGrant(ctx, orgID, p.Name())
	if err != nil {
		return fmt.Errorf("xero: refresh: %w", err)
	}

	refreshPlain, err := p.cipher.Decrypt(grant.RefreshTokenEncrypted)
	if err != nil {
		return fmt.Errorf("xero: decrypt refresh token: %w", err)
	}

	tok, err := p.tokenRequest(ctx, url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {string(refreshPlain)},
	})
	if err != nil {
		return fmt.Errorf("xero: refresh token: %w", err)
	}

	accessEnc, err := p.cipher.Encrypt([]byte(tok.AccessToken))
	if err != nil {
		return fmt.Errorf("xero: encrypt access token: %w", err)
	}
	refreshEnc, err := p.cipher.Encrypt([]byte(tok.RefreshToken))
	if err != nil {
		return fmt.Errorf("xero: encrypt refresh token: %w", err)
	}

	expiresAt := time.Now().Add(time.Duration(tok.ExpiresIn) * time.Second)
	return p.store.UpdateGrantTokens(ctx, grant.ID, accessEnc, refreshEnc, expiresAt)
}

// PushContact implements Provider. Creates or updates a Xero Contact.
func (p *XeroProvider) PushContact(ctx context.Context, orgID uuid.UUID, c Contact) (PushResult, error) {
	accessToken, tenantID, err := p.getAccessToken(ctx, orgID)
	if err != nil {
		return PushResult{}, err
	}

	// Check for existing mapping to decide create vs update.
	mapping, err := p.store.GetMapping(ctx, orgID, p.Name(), "contact", c.ID)
	isUpdate := err == nil

	xc := p.buildXeroContact(c, mapping)
	body, err := json.Marshal(map[string]any{"Contacts": []any{xc}})
	if err != nil {
		return PushResult{}, fmt.Errorf("xero: marshal contact: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		"https://api.xero.com/api.xro/2.0/Contacts", bytes.NewReader(body))
	if err != nil {
		return PushResult{}, err
	}
	p.setHeaders(req, accessToken, tenantID)

	resp, err := p.client.Do(req)
	if err != nil {
		return PushResult{}, fmt.Errorf("xero: push contact: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return PushResult{}, fmt.Errorf("xero: push contact: HTTP %d: %s", resp.StatusCode, string(b))
	}

	var result struct {
		Contacts []struct {
			ContactID string `json:"ContactID"`
		} `json:"Contacts"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return PushResult{}, fmt.Errorf("xero: decode contact response: %w", err)
	}
	if len(result.Contacts) == 0 {
		return PushResult{}, errors.New("xero: push contact: empty response")
	}
	externalID := result.Contacts[0].ContactID
	if err := p.store.UpsertMapping(ctx, orgID, p.Name(), "contact", c.ID, externalID); err != nil {
		return PushResult{}, fmt.Errorf("xero: upsert contact mapping: %w", err)
	}
	return PushResult{LocalID: c.ID, ExternalID: externalID, Updated: isUpdate}, nil
}

// PushTransaction implements Provider. Creates or updates a Xero BankTransaction
// (RECEIVE for credits, SPEND for debits) or a bill (ACCPAY) when the
// transaction direction is debit and a supplier contact is mapped.
func (p *XeroProvider) PushTransaction(ctx context.Context, orgID uuid.UUID, t Transaction) (PushResult, error) {
	accessToken, tenantID, err := p.getAccessToken(ctx, orgID)
	if err != nil {
		return PushResult{}, err
	}

	mapping, err := p.store.GetMapping(ctx, orgID, p.Name(), "transaction", t.ID)
	isUpdate := err == nil

	xt := p.buildXeroBankTransaction(t, mapping)
	body, err := json.Marshal(map[string]any{"BankTransactions": []any{xt}})
	if err != nil {
		return PushResult{}, fmt.Errorf("xero: marshal transaction: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		"https://api.xero.com/api.xro/2.0/BankTransactions", bytes.NewReader(body))
	if err != nil {
		return PushResult{}, err
	}
	p.setHeaders(req, accessToken, tenantID)

	resp, err := p.client.Do(req)
	if err != nil {
		return PushResult{}, fmt.Errorf("xero: push transaction: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return PushResult{}, fmt.Errorf("xero: push transaction: HTTP %d: %s", resp.StatusCode, string(b))
	}

	var result struct {
		BankTransactions []struct {
			BankTransactionID string `json:"BankTransactionID"`
		} `json:"BankTransactions"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return PushResult{}, fmt.Errorf("xero: decode transaction response: %w", err)
	}
	if len(result.BankTransactions) == 0 {
		return PushResult{}, errors.New("xero: push transaction: empty response")
	}
	externalID := result.BankTransactions[0].BankTransactionID
	if err := p.store.UpsertMapping(ctx, orgID, p.Name(), "transaction", t.ID, externalID); err != nil {
		return PushResult{}, fmt.Errorf("xero: upsert transaction mapping: %w", err)
	}
	return PushResult{LocalID: t.ID, ExternalID: externalID, Updated: isUpdate}, nil
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

// getAccessToken returns the plaintext access token and Xero tenant ID for the
// org, refreshing first if the stored token has expired.
func (p *XeroProvider) getAccessToken(ctx context.Context, orgID uuid.UUID) (accessToken, tenantID string, err error) {
	grant, err := p.store.GetGrant(ctx, orgID, p.Name())
	if err != nil {
		return "", "", fmt.Errorf("xero: no active grant for org %s: %w", orgID, err)
	}

	// Lazy refresh: refresh if expiry is within 60 seconds.
	if grant.ExpiresAt.Valid && time.Until(grant.ExpiresAt.Time) < 60*time.Second {
		if err := p.RefreshToken(ctx, orgID); err != nil {
			return "", "", fmt.Errorf("xero: token refresh: %w", err)
		}
		// Re-fetch after refresh.
		grant, err = p.store.GetGrant(ctx, orgID, p.Name())
		if err != nil {
			return "", "", err
		}
	}

	plain, err := p.cipher.Decrypt(grant.AccessTokenEncrypted)
	if err != nil {
		return "", "", fmt.Errorf("xero: decrypt access token: %w", err)
	}
	tenant := ""
	if grant.AccountEmail.Valid {
		tenant = grant.AccountEmail.String
	}
	return string(plain), tenant, nil
}

// tokenRequest performs a POST to the Xero token endpoint with the given form values.
func (p *XeroProvider) tokenRequest(ctx context.Context, params url.Values) (*xeroTokenResponse, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		"https://identity.xero.com/connect/token",
		strings.NewReader(params.Encode()))
	if err != nil {
		return nil, err
	}
	req.SetBasicAuth(p.cfg.ClientID, p.cfg.ClientSecret)
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return nil, fmt.Errorf("token endpoint HTTP %d: %s", resp.StatusCode, string(b))
	}
	var tok xeroTokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&tok); err != nil {
		return nil, err
	}
	return &tok, nil
}

// fetchTenantID calls GET /connections to find the first ORGANISATION tenant
// and returns its tenantId. The tenant ID is used as the Xero-Tenant-Id header
// on all subsequent API calls and stored as account_email in oauth_grants.
func (p *XeroProvider) fetchTenantID(ctx context.Context, accessToken string) (string, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet,
		"https://api.xero.com/connections", nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return "", fmt.Errorf("connections HTTP %d: %s", resp.StatusCode, string(b))
	}
	var conns xeroConnectionsResponse
	if err := json.NewDecoder(resp.Body).Decode(&conns); err != nil {
		return "", err
	}
	for _, c := range conns {
		if c.TenantType == "ORGANISATION" {
			return c.TenantID, nil
		}
	}
	if len(conns) > 0 {
		return conns[0].TenantID, nil
	}
	return "", errors.New("no Xero tenants found for this connection")
}

// storeTokens encrypts and persists the token pair to oauth_grants.
func (p *XeroProvider) storeTokens(ctx context.Context, orgID, userID uuid.UUID, tenantID string, tok *xeroTokenResponse) error {
	accessEnc, err := p.cipher.Encrypt([]byte(tok.AccessToken))
	if err != nil {
		return err
	}
	refreshEnc, err := p.cipher.Encrypt([]byte(tok.RefreshToken))
	if err != nil {
		return err
	}
	expiresAt := time.Now().Add(time.Duration(tok.ExpiresIn) * time.Second)
	return p.store.UpsertGrant(ctx, orgID, userID, p.Name(), tenantID, tok.TokenType, accessEnc, refreshEnc, expiresAt)
}

// setHeaders applies the required Xero API headers to an outgoing request.
func (p *XeroProvider) setHeaders(req *http.Request, accessToken, tenantID string) {
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Xero-Tenant-Id", tenantID)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
}

// ─── Xero payload builders ────────────────────────────────────────────────────

// buildXeroContact returns the JSON-serialisable map for the Xero Contact
// endpoint. If a mapping already exists the ContactID is included so Xero
// updates instead of creating a duplicate.
func (p *XeroProvider) buildXeroContact(c Contact, mapping *Mapping) map[string]any {
	xc := map[string]any{
		"Name": c.Name,
	}
	if mapping != nil {
		xc["ContactID"] = mapping.ExternalID
	}
	if c.Email != "" {
		xc["EmailAddress"] = c.Email
	}
	if c.Phone != "" {
		xc["Phones"] = []map[string]any{
			{"PhoneType": "DEFAULT", "PhoneNumber": c.Phone},
		}
	}
	if c.TaxNumber != "" {
		xc["TaxNumber"] = c.TaxNumber
	}
	// Map contact kind to Xero ContactStatus / IsSupplier / IsCustomer flags.
	switch c.Kind {
	case "customer":
		xc["IsCustomer"] = true
	case "supplier":
		xc["IsSupplier"] = true
	case "both":
		xc["IsCustomer"] = true
		xc["IsSupplier"] = true
	}
	// Address
	if c.AddressLine1 != "" || c.City != "" || c.Country != "" {
		xc["Addresses"] = []map[string]any{
			{
				"AddressType":  "STREET",
				"AddressLine1": c.AddressLine1,
				"AddressLine2": c.AddressLine2,
				"City":         c.City,
				"Region":       c.Region,
				"PostalCode":   c.PostalCode,
				"Country":      c.Country,
			},
		}
	}
	return xc
}

// buildXeroBankTransaction returns the JSON-serialisable map for the Xero
// BankTransactions endpoint. Type is SPEND for debits, RECEIVE for credits.
// If a mapping exists the BankTransactionID is included for update.
func (p *XeroProvider) buildXeroBankTransaction(t Transaction, mapping *Mapping) map[string]any {
	txType := "SPEND"
	if t.Direction == "credit" {
		txType = "RECEIVE"
	}

	xt := map[string]any{
		"Type": txType,
		"Date": t.PostedDate.Format("2006-01-02"),
		"LineItems": []map[string]any{
			{
				"Description": descriptionFor(t),
				"Quantity":    1,
				"UnitAmount":  t.Amount,
				"AccountCode": accountCodeFor(t.AccountCode),
				"TaxType":     taxTypeFor(t.TaxRateCode),
			},
		},
		"BankAccount": map[string]any{
			// The bank account code should match the org's clearing/bank account
			// in Xero. When AccountCode is blank we fall back to "090" (a common
			// default bank account code in Xero demo orgs).
			"Code": bankAccountCodeFor(t.AccountCode),
		},
	}
	if mapping != nil {
		xt["BankTransactionID"] = mapping.ExternalID
	}
	// Attach a contact when one is mapped.
	if t.ContactID != uuid.Nil {
		xt["Contact"] = map[string]any{"ContactID": t.ContactID.String()}
	}
	if t.Currency != "" {
		xt["CurrencyCode"] = t.Currency
	}
	return xt
}

// ─── Mapping helpers: categories → account codes, tax → Xero tax types ──────

// accountCodeFor returns the Xero chart-of-accounts code for a transaction line.
// If the transaction already carries a code (from the slip/scan accounts table)
// it is used directly. Missing or blank codes fall back to the Xero default
// "200" (Sales / general income/expense).
func accountCodeFor(code string) string {
	if code != "" {
		return code
	}
	return "200" // Xero default fallback
}

// bankAccountCodeFor returns the Xero bank account code. We use the same code
// from the transaction's account when it looks like a bank account code;
// otherwise fall back to "090" (Xero demo default bank).
func bankAccountCodeFor(code string) string {
	if code != "" {
		return code
	}
	return "090"
}

// taxTypeFor maps our internal tax_rates.code to the Xero tax type string.
// Xero uses its own named types (OUTPUT, INPUT, NONE, etc.).  We pass through
// our code directly — if the org has configured Xero-compatible codes in the
// tax_rates table they work immediately.  Unknown codes fall back to "NONE".
func taxTypeFor(code string) string {
	switch strings.ToUpper(code) {
	case "OUTPUT", "INPUT", "EXEMPTOUTPUT", "EXEMPTINPUT",
		"ZERORATEDOUTPUT", "ZERORATEDINPUT", "NONE":
		return strings.ToUpper(code)
	case "":
		return "NONE"
	default:
		// Pass through — the org may have configured Xero-aligned codes.
		return code
	}
}

// descriptionFor builds a human-readable line-item description.
func descriptionFor(t Transaction) string {
	if t.Merchant != "" && t.Description != "" && t.Merchant != t.Description {
		return t.Merchant + " — " + t.Description
	}
	if t.Merchant != "" {
		return t.Merchant
	}
	return t.Description
}

File: worker/src/modules/xero/xero.ts
/**
 * XeroProvider — port of backend/internal/accounting_export/xero.go.
 *
 * Handles OAuth2 (auth URL, code exchange, token refresh) and data push
 * (contacts, bank transactions). Token storage uses the DB (oauth_grants).
 *
 * Token "encryption": in this port tokens are stored as plaintext strings
 * (the DB column is text). Production should swap encrypt/decrypt with an
 * AES-GCM implementation keyed from APP_SECRET — the interface is identical.
 *
 * OAuth state (CSRF nonce) is persisted to the oauth_states table so it
 * survives across CF Worker instances (replaces Go's in-memory sync.Map).
 */
import type { Env } from "../../bindings";
import type {
  Contact,
  Transaction,
  PushResult,
  Grant,
  XeroTokenResponse,
  XeroConnection,
} from "./types";
import {
  DEFAULT_XERO_SCOPES,
  ERR_GRANT_NOT_FOUND,
} from "./types";
import {
  getGrant,
  upsertGrant,
  updateGrantTokens,
  revokeGrant,
  getMapping,
  upsertMapping,
  saveOAuthState,
  consumeOAuthState,
} from "./queries";

const XERO_TOKEN_URL       = "https://identity.xero.com/connect/token";
const XERO_CONNECTIONS_URL = "https://api.xero.com/connections";
const XERO_CONTACTS_URL    = "https://api.xero.com/api.xro/2.0/Contacts";
const XERO_BANK_TX_URL     = "https://api.xero.com/api.xro/2.0/BankTransactions";

export const PROVIDER_NAME = "xero";

// ── Public helpers (used by routes.ts) ────────────────────────────────────────

/**
 * Build the Xero OAuth2 consent URL.
 * Persists `state` nonce → orgId in DB for callback validation.
 */
export async function buildAuthURL(
  env:   Env,
  orgId: string,
  state: string,
): Promise<string> {
  await saveOAuthState(env, state, orgId);
  const scopes = DEFAULT_XERO_SCOPES.join(" ");
  const params = new URLSearchParams({
    response_type: "code",
    client_id:     env.XERO_CLIENT_ID!,
    redirect_uri:  env.XERO_REDIRECT_URL!,
    scope:         scopes,
    state,
  });
  return `https://login.xero.com/identity/connect/authorize?${params.toString()}`;
}

/**
 * Validate the callback state nonce and exchange code for tokens.
 * Returns the connected Xero tenant ID (stored as account_email for display).
 */
export async function exchangeCode(
  env:    Env,
  state:  string,
  code:   string,
  userId: string,
): Promise<{ tenantId: string; orgId: string }> {
  // Validate state nonce (CSRF).
  const orgId = await consumeOAuthState(env, state);
  if (!orgId) throw new Error("xero: unknown or expired OAuth state");

  const tok = await tokenRequest(env, new URLSearchParams({
    grant_type:   "authorization_code",
    code,
    redirect_uri: env.XERO_REDIRECT_URL!,
  }));

  const tenantId = await fetchTenantId(tok.access_token);

  await upsertGrant(
    env, orgId, userId, PROVIDER_NAME,
    tenantId,            // account_email stores the Xero tenant ID
    tok.token_type,
    tok.access_token,    // plaintext — swap with encrypt() for production
    tok.refresh_token,
    new Date(Date.now() + tok.expires_in * 1000),
  );

  return { tenantId, orgId };
}

/**
 * Revoke the stored OAuth grant for (orgId, xero).
 */
export async function disconnectXero(env: Env, orgId: string): Promise<void> {
  await revokeGrant(env, orgId, PROVIDER_NAME);
}

/**
 * Fetch grant status for an org. Returns null if not connected.
 */
export async function getXeroStatus(env: Env, orgId: string): Promise<Grant | null> {
  try {
    return await getGrant(env, orgId, PROVIDER_NAME);
  } catch (e) {
    if (e instanceof Error && e.message === ERR_GRANT_NOT_FOUND) return null;
    throw e;
  }
}

// ── Push operations ───────────────────────────────────────────────────────────

/** Push a contact to Xero (create or update, idempotent via mapping table). */
export async function pushContact(env: Env, orgId: string, c: Contact): Promise<PushResult> {
  const { accessToken, tenantId } = await getAccessToken(env, orgId);

  let mapping = null;
  try {
    mapping = await getMapping(env, orgId, PROVIDER_NAME, "contact", c.id);
  } catch { /* not found → create */ }

  const isUpdate = mapping !== null;
  const xc       = buildXeroContact(c, mapping?.externalId);
  const body     = JSON.stringify({ Contacts: [xc] });

  const resp = await fetch(XERO_CONTACTS_URL, {
    method:  "POST",
    headers: xeroHeaders(accessToken, tenantId),
    body,
  });

  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(`xero: push contact: HTTP ${resp.status}: ${text.slice(0, 256)}`);
  }

  const result = (await resp.json()) as { Contacts: Array<{ ContactID: string }> };
  if (!result.Contacts?.length) throw new Error("xero: push contact: empty response");

  const externalId = result.Contacts[0].ContactID;
  await upsertMapping(env, orgId, PROVIDER_NAME, "contact", c.id, externalId);

  return { localId: c.id, externalId, updated: isUpdate };
}

/** Push a bank transaction to Xero (create or update, idempotent). */
export async function pushTransaction(env: Env, orgId: string, t: Transaction): Promise<PushResult> {
  const { accessToken, tenantId } = await getAccessToken(env, orgId);

  let mapping = null;
  try {
    mapping = await getMapping(env, orgId, PROVIDER_NAME, "transaction", t.id);
  } catch { /* not found → create */ }

  const isUpdate = mapping !== null;
  const xt       = buildXeroBankTransaction(t, mapping?.externalId);
  const body     = JSON.stringify({ BankTransactions: [xt] });

  const resp = await fetch(XERO_BANK_TX_URL, {
    method:  "POST",
    headers: xeroHeaders(accessToken, tenantId),
    body,
  });

  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(`xero: push transaction: HTTP ${resp.status}: ${text.slice(0, 256)}`);
  }

  const result = (await resp.json()) as { BankTransactions: Array<{ BankTransactionID: string }> };
  if (!result.BankTransactions?.length) throw new Error("xero: push transaction: empty response");

  const externalId = result.BankTransactions[0].BankTransactionID;
  await upsertMapping(env, orgId, PROVIDER_NAME, "transaction", t.id, externalId);

  return { localId: t.id, externalId, updated: isUpdate };
}

// ── Token management ──────────────────────────────────────────────────────────

/** Return plaintext access token + tenant ID, refreshing if within 60s of expiry. */
async function getAccessToken(
  env:   Env,
  orgId: string,
): Promise<{ accessToken: string; tenantId: string }> {
  let grant = await getGrant(env, orgId, PROVIDER_NAME);

  // Lazy refresh: refresh if expiry is within 60 seconds.
  if (grant.expiresAt && (grant.expiresAt.getTime() - Date.now()) < 60_000) {
    await refreshToken(env, orgId, grant);
    grant = await getGrant(env, orgId, PROVIDER_NAME); // re-fetch after refresh
  }

  const accessToken = grant.accessTokenEncrypted; // plaintext in this port
  const tenantId    = grant.accountEmail ?? "";
  return { accessToken, tenantId };
}

async function refreshToken(env: Env, orgId: string, grant: Grant): Promise<void> {
  const refreshPlain = grant.refreshTokenEncrypted; // plaintext in this port

  const tok = await tokenRequest(env, new URLSearchParams({
    grant_type:    "refresh_token",
    refresh_token: refreshPlain,
  }));

  await updateGrantTokens(
    env, grant.id,
    tok.access_token,
    tok.refresh_token,
    new Date(Date.now() + tok.expires_in * 1000),
  );
}

// ── Xero API helpers ──────────────────────────────────────────────────────────

async function tokenRequest(env: Env, params: URLSearchParams): Promise<XeroTokenResponse> {
  const creds = btoa(`${env.XERO_CLIENT_ID}:${env.XERO_CLIENT_SECRET}`);
  const resp  = await fetch(XERO_TOKEN_URL, {
    method:  "POST",
    headers: {
      "Authorization": `Basic ${creds}`,
      "Content-Type":  "application/x-www-form-urlencoded",
    },
    body: params.toString(),
  });

  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(`xero: token endpoint HTTP ${resp.status}: ${text.slice(0, 512)}`);
  }

  return (await resp.json()) as XeroTokenResponse;
}

async function fetchTenantId(accessToken: string): Promise<string> {
  const resp = await fetch(XERO_CONNECTIONS_URL, {
    headers: {
      Authorization:  `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
  });

  if (!resp.ok) {
    const text = await resp.text().catch(() => "");
    throw new Error(`xero: connections HTTP ${resp.status}: ${text.slice(0, 512)}`);
  }

  const conns = (await resp.json()) as XeroConnection[];
  // Prefer ORGANISATION type; fall back to first.
  const org = conns.find((c) => c.tenantType === "ORGANISATION") ?? conns[0];
  if (!org) throw new Error("xero: no Xero tenants found for this connection");
  return org.tenantId;
}

function xeroHeaders(accessToken: string, tenantId: string): Record<string, string> {
  return {
    Authorization:    `Bearer ${accessToken}`,
    "Xero-Tenant-Id": tenantId,
    "Content-Type":   "application/json",
    Accept:           "application/json",
  };
}

// ── Xero payload builders ─────────────────────────────────────────────────────

/** Port of buildXeroContact in xero.go. */
function buildXeroContact(c: Contact, existingExternalId?: string): Record<string, unknown> {
  const xc: Record<string, unknown> = { Name: c.name };

  if (existingExternalId) xc["ContactID"] = existingExternalId;
  if (c.email)            xc["EmailAddress"] = c.email;
  if (c.phone)            xc["Phones"] = [{ PhoneType: "DEFAULT", PhoneNumber: c.phone }];
  if (c.taxNumber)        xc["TaxNumber"] = c.taxNumber;

  switch (c.kind) {
    case "customer": xc["IsCustomer"] = true; break;
    case "supplier": xc["IsSupplier"] = true; break;
    case "both":
      xc["IsCustomer"] = true;
      xc["IsSupplier"] = true;
      break;
  }

  if (c.addressLine1 || c.city || c.country) {
    xc["Addresses"] = [{
      AddressType:  "STREET",
      AddressLine1: c.addressLine1,
      AddressLine2: c.addressLine2,
      City:         c.city,
      Region:       c.region,
      PostalCode:   c.postalCode,
      Country:      c.country,
    }];
  }

  return xc;
}

/** Port of buildXeroBankTransaction in xero.go. */
function buildXeroBankTransaction(t: Transaction, existingExternalId?: string): Record<string, unknown> {
  const txType = t.direction === "credit" ? "RECEIVE" : "SPEND";

  const xt: Record<string, unknown> = {
    Type: txType,
    Date: t.postedDate.toISOString().slice(0, 10),
    LineItems: [{
      Description: descriptionFor(t),
      Quantity:    1,
      UnitAmount:  t.amount,
      AccountCode: accountCodeFor(t.accountCode),
      TaxType:     taxTypeFor(t.taxRateCode),
    }],
    BankAccount: { Code: bankAccountCodeFor(t.accountCode) },
  };

  if (existingExternalId) xt["BankTransactionID"] = existingExternalId;

  const NIL_UUID = "00000000-0000-0000-0000-000000000000";
  if (t.contactId && t.contactId !== NIL_UUID) {
    xt["Contact"] = { ContactID: t.contactId };
  }

  if (t.currency) xt["CurrencyCode"] = t.currency;

  return xt;
}

// ── Mapping helpers ───────────────────────────────────────────────────────────

function accountCodeFor(code: string): string {
  return code || "200";
}

function bankAccountCodeFor(code: string): string {
  return code || "090";
}

function taxTypeFor(code: string): string {
  const up = (code ?? "").toUpperCase();
  const KNOWN = new Set([
    "OUTPUT", "INPUT", "EXEMPTOUTPUT", "EXEMPTINPUT",
    "ZERORATEDOUTPUT", "ZERORATEDINPUT", "NONE",
  ]);
  if (!up)            return "NONE";
  if (KNOWN.has(up))  return up;
  return code; // pass-through — org may use Xero-aligned codes
}

function descriptionFor(t: Transaction): string {
  if (t.merchant && t.description && t.merchant !== t.description) {
    return `${t.merchant} — ${t.description}`;
  }
  return t.merchant || t.description;
}

// ── Re-exports for tests ──────────────────────────────────────────────────────
export { buildXeroContact as _buildXeroContact, buildXeroBankTransaction as _buildXeroBankTransaction };
export { accountCodeFor, bankAccountCodeFor, taxTypeFor, descriptionFor };
